cloudinit/net/dhcp.py

```diff
@@ -281,24 +304,30 @@ def parse_static_routes(rfc3442):
     """ parse rfc3442 format and return a list containing tuple of strings.
 
     The tuple is composed of the network_address (including net length) and
-    gateway for a parsed static route.
+    gateway for a parsed static route.  It can parse two formats of rfc3442,
+    one from dhcpcd and one from dhclient (isc).
 
-    @param rfc3442: string in rfc3442 format
+    @param rfc3442: string in rfc3442 format (isc or dhcpd)
     @returns: list of tuple(str, str) for all valid parsed routes until the
               first parsing error.
 
     E.g.
-    sr = parse_state_routes("32,169,254,169,254,130,56,248,255,0,130,56,240,1")
-    sr = [
+    sr=parse_static_routes("32,169,254,169,254,130,56,248,255,0,130,56,240,1")
+    sr=[
         ("169.254.169.254/32", "130.56.248.255"), ("0.0.0.0/0", "130.56.240.1")
     ]
 
+    sr2 = parse_static_routes("24.191.168.128 192.168.128.1,0 192.168.128.1")
+    sr2 = [
+        ("191.168.128.0/24", "192.168.128.1"), ("0.0.0.0/0", "192.168.128.1")
+    ]
+
     Python version of isc-dhclient's hooks:
        /etc/dhcp/dhclient-exit-hooks.d/rfc3442-classless-routes
     """
     # raw strings from dhcp lease may end in semi-colon
     rfc3442 = rfc3442.rstrip(";")
-    tokens = rfc3442.split(',')
+    tokens = list(filter(None, re.split(r"[, .]", rfc3442)))
```

Comment:
```suggestion
    tokens = [tok for tok in re.split(r"[, .]", rfc3442) if tok]
```

Reply:
I do not like this generator expression over the filter call.  I prefer to be explicit in what I'm trying to do.  In this case, the filter to None actually remove all empty field (including those that would evaluate to false like "").  

If you're adamant on wanting to use the generator expression, then my proposition would be to put it in a different function having a better name and then I would have the generator expression in the function.

Reply:
so something like this could be use to hold the generator expression;
`
def extract_all_relevant_tokens(rfc3442):
    return list(filter(None, re.split(r"[, .]", rfc3442)))`

Reply:
The list comprehension is explicit, to me, because it doesn't rely on the reader knowing the implicit behaviour of `filter` when `None` is passed as the first argument. (As a data point, I've been writing Python for years and I had to look up what `None` as first argument to filter does.)

From a consistency point of view, we currently only have 2 calls to `filter` in the entire cloud-init codebase, and over 100 instances of _single-line_ list comprehensions (`r"\[.*for.*in"`).

(As a side note, generator expressions are contained in parentheses (`(tok for tok in ... if tok)`) and produce a generator (as `filter` does). If a generator expression were used, it would also require a call to `list`.)

Reply:
It still feels wrong to me, but I guess I'll have to deal with it :).  Let me know if anything else :)

Thanks @OddBloke, your arguments made sense to me.